### ISSUE TITLE
Lookup types and polymorphic this issue

### ISSUE THREAD

**Issue by dsehnal** [Microsoft/TypeScript#12723]
**TypeScript Version:**  2.1.4

**Code**

```ts
interface A<T> { value: T }
interface B extends A<string> { }

abstract class C<T extends A<any>> {
    private t: T;
    get a(): this['t']['value'] { return this.t.value };
}

class D extends C<B> {
    get b() { return this.a; }
}

```

**Expected behavior:**

I would expect that the type of ``D.b`` would be ``string``.

**Actual behavior:**

Type of  ``D.b`` is ``any``

**Comment by dsehnal** [comment]
Here is a more simplified version of the issue:

```TypeScript
interface A<T> { t: T }
function f<T extends A<any>>(t: T['t']): T {
    return { t };
}
```

Here ``t`` is of type ``any`` instead of ``T['t']``. So I suppose that what's going on is that ``T['t']`` is evaluated immediately instead of keeping the generic value.

So my question would be: Will the type system ever support this behaviour?

**Comment by mhegazy** [comment]
This is a bug, should be fixed shortly.

**Comment by ahejlsberg** [comment]
The issue is fixed by #12770, but note that your example only works if you remove the `private` modifier from `t` because indexed access types only operate on public properties.
